### PR TITLE
WHy use typedresults /3

### LINKED ISSUE #28850: Provide explanation for TypedResults
It is unclear what TypedResults provide here. There should be either be an explanation of why to use them or a link to a page where it is explained.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: ab408035-6313-a183-9940-32de00e0e977
* Version Independent ID: ab408035-6313-a183-9940-32de00e0e977
* Content: [Tutorial: Create a minimal API with ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?view=aspnetcore-7.0&tabs=visual-studio)
* Content Source: [aspnetcore/tutorials/min-web-api.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/tutorials/min-web-api.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-tutorials**
* GitHub Login: @wadepickett
* Microsoft Alias: **wpickett**

---
[Associated WorkItem - 85494](https://dev.azure.com/msft-skilling/Content/_workitems/edit/85494)

### LINKED ISSUE #28917: Move Built-in results table to API link
Remove the table at [Built-in results](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0#built-in-results) to a link to [Results class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.results?view=aspnetcore-7.0)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: b09824ab-b051-75a1-13b7-b0e281cc46e0
* Version Independent ID: b09824ab-b051-75a1-13b7-b0e281cc46e0
* Content: [Minimal APIs quick reference](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0)
* Content Source: [aspnetcore/fundamentals/minimal-apis.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/minimal-apis.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 85495](https://dev.azure.com/msft-skilling/Content/_workitems/edit/85495)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #28850
Fixes #28917

Add code from https://github.com/dotnet/AspNetCore.Docs.Samples/pull/140/files

Better preview URLs that go to the correct place
* [typedresults-vs-results](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/responses?branch=pr-en-us-28918&view=aspnetcore-7.0#typedresults-vs-results)
* [Use the TypedResults API](https://review.learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?branch=pr-en-us-28918&view=aspnetcore-7.0&tabs=visual-studio#use-the-typedresults-api)
* [aspnetcore/fundamentals/minimal-apis](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?branch=pr-en-us-28918#return-typedresults)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/minimal-apis.md](https://github.com/dotnet/AspNetCore.Docs/blob/bd2b80e6ac00017f6551f0f1f39ba0a95e02bc71/aspnetcore/fundamentals/minimal-apis.md) | [Differences between minimal APIs and APIs with controllers](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?branch=pr-en-us-28918) |
| [aspnetcore/fundamentals/minimal-apis/responses.md](https://github.com/dotnet/AspNetCore.Docs/blob/bd2b80e6ac00017f6551f0f1f39ba0a95e02bc71/aspnetcore/fundamentals/minimal-apis/responses.md) | [How to create responses in Minimal API apps](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/responses?branch=pr-en-us-28918) |
| [aspnetcore/tutorials/min-web-api.md](https://github.com/dotnet/AspNetCore.Docs/blob/bd2b80e6ac00017f6551f0f1f39ba0a95e02bc71/aspnetcore/tutorials/min-web-api.md) | [[Visual Studio for Mac](#tab/visual-studio-mac)](https://review.learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?branch=pr-en-us-28918) |

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@JeremyLikness maybe you can recommend the right reviewer.

**Comment by davidfowl** [comment]
cc @DamianEdwards

File: aspnetcore/fundamentals/minimal-apis/responses.md [hunk 1]
```diff
@@ -81,6 +81,39 @@ app.MapGet("/hello", () => TypedResults.Ok(new Message() {  Text = "Hello World!
 
 For more information about describing a response type, see [OpenAPI support in minimal APIs](/aspnet/core/fundamentals/minimal-apis/openapi#describe-response-types-1).
 
+As mentioned previously, when using `TypedResults`, a conversion is not needed. Consider the following minimal API which returns a `TypedResults` class
+
+<!-- TODO, https://github.com/dotnet/AspNetCore.Docs/issues/28916 moves the following code to a project -->
+
+```csharp
+public static async Task<IResult> GetAllTodos(TodoDb db)
```
Comment (DamianEdwards) [inline aspnetcore/fundamentals/minimal-apis/responses.md:89]:
```suggestion
public static async Task<Ok<Todo[]>> GetAllTodos(TodoDb db)
```

Note that this example works the same with `Results.Ok()` as even though the method returns `IResult`, the concrete type can still be checked the same way in the test.

**Comment by DamianEdwards** [comment]
The other thing to consider adding is that because all methods on `Results` return `IResult` in their signature, the compiler can automatically infer that as the request delegate return type when returning different results from a single endpoint, whereas `TypedResults` requires the use of `Results<T1, TN>` from such delegates, e.g.:

This method compiles because both `Results.Ok` and `Results.NotFound` are declared as returning `IResult`, even though the actual concrete types of the objects returned are different:

```csharp
app.MapGet("/todos/{id}", async (int id, Db db) =>
    await db.FirstOrDefaultAsync(id) is { } todo
        ? Results.Ok(todo)
        : Results.NotFound());
```

The following method does not compile, because `TypedResults.Ok` and `TypedResults.NotFound` are declared as returning different types and the compiler will not attempt to infer the "best matching" type:

```csharp
app.MapGet("/todos/{id}", async (int id, Db db) =>
    await db.FirstOrDefaultAsync(id) is { } todo
        ? TypedResults.Ok(todo)
        : TypedResults.NotFound());
```

To use `TypedResults`, the return type must be fully declared, which when async also necessitates the `Task<>` wrapper. So it's quite a bit more verbose, but that's the trade-off for having the type information be statically available and thus capable of self-describing to OpenAPI:

```csharp
app.MapGet("/todos/{id}", async Task<Results<Ok<Todo>, NotFound>> (int id, Db db) =>
    await db.FirstOrDefaultAsync(id) is { } todo
        ? TypedResults.Ok(todo)
        : TypedResults.NotFound());
```

**Review by captainsafia — APPROVED** [review]

**Pull request merged**